### PR TITLE
Add first draft of interpreter module

### PR DESCRIPTION
**Pull request by dannywillems**
Based on @mrmr1993's work. Importing from https://github.com/o1-labs/proof-systems/pull/1332

### CODE REVIEW

File: optimism/src/mips/interpreter.rs [hunk 1]
```diff
@@ -55,9 +59,9 @@ pub enum RTypeInstruction {
     Nor,                          // nor
     SetLessThan,                  // slt
     SetLessThanUnsigned,          // sltu
-    MultiplyToRegister,           // mul
-    CountLeadingOnes,             // clo
-    CountLeadingZeros,            // clz
+    MultiplyToRegister,           // mul - FIXME: pseudo?
+    CountLeadingOnes,             // clo - FIXME: pseudo?
+    CountLeadingZeros,            // clz - FIXME: pseudo?
```
Comment (rbonichon) [inline optimism/src/mips/interpreter.rs:64]: I don't quite get the "pseudo" comment here. All these instructions are documented in the official instruction set description from [here](https://mips.com/products/architectures/mips32-2/)

File: optimism/src/mips/interpreter.rs [hunk 2]
```diff
@@ -34,9 +38,9 @@ pub enum RTypeInstruction {
     SyscallWriteOther,            // syscall (Write ?)
     SyscallFcntl,                 // syscall (Fcntl)
     SyscallOther,                 // syscall (Brk, Clone, ?)
-    MoveZero,                     // movz
-    MoveNonZero,                  // movn
-    Sync,                         // sync
+    MoveZero,                     // movz - FIXME: pseudo?
```
Comment (rbonichon) [inline optimism/src/mips/interpreter.rs:41]: `movz` is documented as "MIPS32 removed in  Release 6"

File: optimism/src/mips/interpreter.rs [hunk 3]
```diff
@@ -34,9 +38,9 @@ pub enum RTypeInstruction {
...
+    MoveNonZero,                  // movn - FIXME: pseudo?
```
Comment (rbonichon) [inline optimism/src/mips/interpreter.rs:42]: `movn` marked as "removed in Release 6" too

File: optimism/src/mips/interpreter.rs [hunk 4]
```diff
@@ -34,9 +38,9 @@ pub enum RTypeInstruction {
...
+    Sync,                         // sync - FIXME: pseudo?
```
Comment (rbonichon) [inline optimism/src/mips/interpreter.rs:43]: `sync` is documented in its own right in the manual. Doesn't look "pseudo" to me

File: optimism/src/mips/interpreter.rs [hunk 5]
```diff
@@ -93,3 +98,899 @@ pub enum ITypeInstruction {
...
+                .map(&Instruction::RType)
+                .chain(JTypeInstruction::iter().map(&Instruction::JType))
+                .chain(ITypeInstruction::iter().map(&Instruction::IType)),
+        )
+    }
+}
+
+// pub fn all_instruction_selectors() -> impl Iterator<Item = Column> {
+//     Instruction::iter().map(Column::Instruction)
+// }
```
Comment (rbonichon) [inline optimism/src/mips/interpreter.rs:893]: Delete?

**Review by rbonichon — APPROVED** [review]

File: optimism/src/mips/interpreter.rs (hunk 2 above)
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs]: Added as a comment in the code.

File: optimism/src/mips/interpreter.rs (hunk 1 above)
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs]: Removed

File: optimism/src/mips/interpreter.rs (hunk 3 above)
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs]: Added as a comment in the code.

**Pull request closed without merging**